src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs

```diff
@@ -13,6 +13,46 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
 {
     public sealed partial class HttpRequestHeaders : HttpHeaders
     {
+        private readonly bool _reuseHeaderValues;
+        private long _previousBits = 0;
+
+        public HttpRequestHeaders(bool reuseHeaderValues = false)
+        {
+            _reuseHeaderValues = reuseHeaderValues;
+        }
+
+        public void OnHeadersComplete()
+        {
+            var bitsToClear = _previousBits & ~_bits;
+            _previousBits = 0;
+
+            if (bitsToClear != 0)
+            {
+                // Some previous headers were not reused or overwritten, so clear them now.
+                Clear(bitsToClear);
+            }
+        }
+
+        protected override void ClearFast()
+        {
+            if (!_reuseHeaderValues)
+            {
+                // If we aren't reusing headers clear them all
+                Clear(_bits);
+            }
+            else
+            {
+                // If we are reusing headers, store the currently set headers for comparision later
```

Comment:
nit: spelling "comparision" => "comparison"
```suggestion
                // If we are reusing headers, store the currently set headers for comparison later
```

Reply:
Done